### DISCUSSION TITLE
[Information] Got my account suspended

### DISCUSSION THREAD

**Discussion by user234683** [victornpb/undiscord#273] [2. Discussions]
I began deleting all my account's messages 2 days ago. Probably somewhere around 10,000 in total. With an average delay of 2450 ms (I used the modification #168 which adjusts the delay randomly, however, I applied the changes from that gist to the original script and also set it to increase the min and max delays if rate limiting is reached). However, I didn't ever notice rate limits being reached. I used a min of 900 ms and a max of 4000 ms.

After getting most of it deleted, what seems to have resulted in my account suspended was one of or a combination of the following factors:

(1) Had the third party client Ripcord open
(2) Sent a message through it to a non-friend (I thought he was on my friends list but turns out no). Message [failed to send](https://support.discord.com/hc/en-us/articles/360060145013).
(3) Then I sent a friend request. Around this point I was logged out and my account was disabled.

So if you're going to do mass-deleting, do not do any of those three things!

**Comment by FunnyInternet** [comment]
Deleted tens of thousands of my messages and also have used ripcord for a long time and my account is fine, so it's either completely random or you did something else? Did discord explain the reason at all in an email or something

**Comment by cedws** [comment]
I am a maintainer of another deletion tool and have had one of my own testing accounts banned today - the first in two or three years I've been working on it.

I suspect Discord are cracking down on self-botting. I will be adding more warnings to my tool and I strongly suggest some warnings are added to this tool also.

If you're a user, you should back up anything you don't want to lose. I doubt Discord will accept appeals.

**Comment by victornpb** [comment]
@cedws I will probably add a banner on the project readme and inside the UI.

This project really started as a personal scripts, but since more and more (non tech) people are using it, it makes sense to make them aware or the risks.

I strongly discourage deleting messages other than your own, unless its a handful of messages. If you need to delete a large amount of messages, don't push it very aggressively, you don't want to bring attention by hammering the discord API.

**Reply by Xotywod** [reply]
what's ways to avoid getting ur account suspended and banned if you could give me a list and is there any better delete delays time intervals that would be less harsh on the discord app?

**Comment by victornpb** [comment]
I added a warning to the README, and added a link to this thread. This issue is also pinned.
I will also add a warning inside the UI later (I just don't have time to change the HTML+CSS right now).

**Comment by cedws** [comment]
Nice one.

One way to get around the self-botting rules may be to automate actions such as clicks and keypresses via the Discord client so that it looks legitimate.

What I will probably aim to find out is the heuristics Discord are using on the backend to detect self-botting. I hope they are using a simple set of rules rather than some black box machine learning. Some ideas:

* Detection of high volume of requests or 429s
* Detection of hitting specific endpoint too often
* Detection of requests without corresponding `/science` requests (telemetry endpoint)
* Detection of requests without specific headers

I'll let you know if I find anything.

**Comment by githubuser843** [comment]
Seems to work fine in small message amounts even on fresh account so not sure what specifically is causing it. Are there any specific guidelines on using it and not having your account removed? Is it just to use it normally and not DM/use other discord clients

**Comment by user234683** [comment]
I believe what causes your account to get removed is not the deleting itself, but doing mass deleting and then doing any of the following actions:

(1) New friend requests
(2) Sending messages
(3) Joining servers

The reason is that Discord's automated system is trying to crack down on automated accounts that add people as friends or join servers and then send spam, scams, and other automated junk to them. So I think mass deleting lowers your "account reputation," and results in your account being noted as engaged in self-botting, but it really only triggers the anti-spam system/suspension in those 3 cases. So I would advise if you're mass deleting, do not send anyone DMs, and do not add anyone as friends or join any servers. Only delete, and then wait maybe a couple days before doing anything else so your account reputation can reset. This is just speculation though.

**Comment by Vortex545** [comment]
> I am a maintainer of another deletion tool and have had one of my own testing accounts banned today - the first in two or three years I've been working on it.
> 
> I suspect Discord are cracking down on self-botting. I will be adding more warnings to my tool and I strongly suggest some warnings are added to this tool also.
> 
> If you're a user, you should back up anything you don't want to lose. I doubt Discord will accept appeals.

I got suspended and i wasn't even using the bot, just deleting manually. Very strange

**Comment by MkUltraCIA** [comment]
> I got suspended and i wasn't even using the bot, just deleting manually. Very strange

That's even worse.

**Comment by Animan8000** [comment]
Were the affected accounts by any chance unverified Discord accounts? Because I am aware of some Anti-Spam system that kicks in easily on unverified ones in certain situations, forcing you to provide email or phone number to continue.

**Comment by user234683** [comment]
My account was verified. I had run into the anti-spam system unintentionally before, because I asked a similar question in two servers, and then joined a new one, after which it made me give a phone number.

**Comment by Falcc** [comment]
Just found this, seems to be pretty clear:
https://support.discord.com/hc/en-us/articles/115002192352-Automated-user-accounts-self-bots-

`Automating normal user accounts (generally called "self-bots") outside of the OAuth2/bot API is forbidden, and can result in an account termination if found.`

**Comment by githubuser843** [comment]
Any update on any of this? Not sure I want to have my account banned for deleting messages. If it happens in the middle of it then you can't really remove the rest. I'm not really sure what makes it incredibly noticeable, is it just a large volume is suspicious? Stretching the limits of the discord API or whatever.

**Comment by Animan8000** [comment]
I think the very minimum you can do (while keep using it) is not to stress the API additionally. Which means don't look into other channels, don't send messages, don't send friend requests, etc. while the script is doing it's job. Optionally changing the delay can be done but that is up to you.

**Comment by Falcc** [comment]
I was able to run this a few days ago, and haven't been banned yet. But I modified it to be much less brutal on the API.

I changed the "baseDeleteDelay" variable to 700 on line 151.
And I added a random number between 1 and 100 to the delete delay on line 335.

I did this because I found that a baseDeleteDelay (and no random number) of 750 was the sweet spot for never getting rate limited. Dropping it to 700 and adding the rand of 1 to 100 should give me an average of 750, but make it look (slightly) less automated.

I deleted approx 35000 messages in a single DM over about 10 hours.

**Comment by adnit** [comment]
> I was able to run this a few days ago, and haven't been banned yet. But I modified it to be much less brutal on the API.
> 
> I changed the "baseDeleteDelay" variable to 700 on line 151.
> And I added a random number between 1 and 100 to the delete delay on line 335.
> 
> I did this because I found that a baseDeleteDelay (and no random number) of 750 was the sweet spot for never getting rate limited. Dropping it to 700 and adding the rand of 1 to 100 should give me an average of 750, but make it look (slightly) less automated.
> 
> I deleted approx 35000 messages in a single DM over about 10 hours.

this seems to work for me too, thanks for the idea @m0illi

**Comment by Arcitec** [comment]
@m0illi I see that you are using an old version of the script. The latest version from March 30th 2021 has already changed the default deletion delay to 1 second, and also doesn't have the variables you talk about. It also now dynamically extends the delay if you get rate limited.

If you wanna have a random deletion delay in the new version, I narrowed it down to the line that says:

```js
await wait(deleteDelay);
```

**But that's not the best place to add the delay.** There's many different API calls that use the wait-function. So instead, search for this line (for me it was near the absolute top of the file):

```js
const wait = async ms => new Promise(done => setTimeout(done, ms));
```

And change it to this (randomly adds 0-200ms):

```js
const wait = async ms => new Promise(done => setTimeout(done, ms + Math.floor(Math.random() * 200) ));
```

This will make every API delay slightly random. It will also mean that the "estimated time remaining" is no longer true in the GUI but that should be obvious.

Anyway I dunno if this will help. It might. It's possible that they look for a pattern of API requests coming with the exact same timings since the previous request. But honestly, it's even more likely that they simply look for fast requests in general, and lack of things like intermittent "normal user" API requests inbetween.

But having a random delay can't hurt, so I did it... I also manually (in the GUI part of this script after clicking the trash can icon) set my core deletion delay to 1500 (instead of 1000) and my search delay to 500 (instead of 100) because I am not in a rush and I don't want to get banned.

**Comment by SwampyApple** [comment]
@Bananaman thank you for the addition. Have you had any further results? I haven't been seeing any recent comments on suspensions in this issue but that's no guarantee that anything has changed meanwhile. I also think it would be valuable to know the exact given reason and duration of the suspension/ban.

**Comment by Falcc** [comment]
I haven't been banned using the 700ms + rand(0,100). I've used it multiple
times over the last few weeks. First time running it was ran for over 9
hours, same with second time. Then a few more times less than 15 Min each.

**Comment by SwampyApple** [comment]
> 
> I haven't been banned using the 700ms + rand(0,100). I've used it multiple times over the last few weeks. First time running it was ran for over 9 hours, same with second time. Then a few more times less than 15 Min each.

Thanks for the additional information. One extra question: was your account already phone verified? I've been reading that some users had their accounts locked out until they verified through their phone.

**Comment by Arcitec** [comment]
@SwampyApple @m0illi I used the settings described earlier:

- changed deletion delay to 1500 (instead of 1000) and my search delay to 500 (instead of 100)
- my patch to add 0-200ms random extra delay to each API call.
- I do not have phone verification. I do not have a phone number added. I've never had a phone number associated with it. I only have a random protonmail email.
- I deleted 3000 messages all at once with these settings.
- 11 days have passed after that and I am not banned! Edit: A few months later. Not banned.

**Comment by AllEternalsDeck** [comment]
I mass deleted about 10000 messages in multiple channels with the default settings (1000s), then changed my mail on my account, then created a new account with the old mail, then set the old account for deletion. Do you reckon I'm at risk for any ban (especially with the new account?)

Sorry this might not be pertinent to the specifics of the topic, but I have anxiety issues, and getting my account banned is nagging at my head very heavily right now.

**Comment by cedws** [comment]
@AllEternalsDeck There is no certainty unless you work at Discord and know the botting heuristics they have in place. If being banned is a problem for you, you need to assume it will happen and plan accordingly.

**Comment by adnit** [comment]
@AllEternalsDeck probably not since most probably accounts are tied to a userId and email is not fully taken into consideration, for ex if u were indeed about to get banned on ur old account and changing the email address would have prevented the ban

**Comment by AllEternalsDeck** [comment]
If I may ask how long after the deleting of the messages have people been banned?

Also are the bans you've been receiving temporary suspensions, account disabling or straight up IP bans?

I am very worried about IP banning because discord's stance on the matter doesn't really specify the kind of banning associated with self botting...

**Comment by DAKiersz** [comment]
Last week, I had a chance to test the script with trivial modifications as suggested by @Bananaman. I've conducted a large-scale test on two accounts that resulted in the deletion of 80k and 20k messages respectively in groups, DMs etc. So far, no ban or bot detection mechanism has been triggered. Here, I want to list some of the paranoid precautions that were done:

* Similarly to @Bananaman, I've changed to default values of deleteDelay to 1500ms and searchDelay to 1000ms).
* A random, additional delay to the API call was set to 20% of the deleteDelay - a bit arbitrary but it should generalize better for different values of deleteDelay.
* On both accounts, I've disconnected the mobile phone authentication and changed both emails to throwaways, again, to recreate any accounts that may perish. 
* Prior to logging in to Discord using Chromium (via two different devices for each account), I logged out of all devices - this can be done by simply changing your password and logging in again. 
* A bit of a long shot, but using a VPN during these deletion sessions (and in general!) should not hurt. I've done this to see if a ban would target the IP. This is probably the most speculative point on this list.
* The scripts ran for about just over 2 days on the first account with no interruptions (say, to the API limiter).
* Going along with @user234683 suggestion, accounts were 'let be' for about three days after mass deletion - particularly when it comes to DMing anyone. It does seem logical that any anti-spambot system may see sending DMs as problematic. There is of course no evidence for any of this. 
* Lastly, I suggest changing your password and 2FA to reset authToken afterwards just in case a third party somehow intercepted this string.

**None of the above** are a known silver bullet to this issue and are, at best, **educational guesses.** Without knowing the inner workings of the bot-detection system, we can only attempt to confuse such mechanisms, as rightly suggested by @cedws.

@AllEternalsDeck As mentioned above, there will never be any guarantee that you would _not_ get banned. Plan accordingly, undertake preparations and assume that it _may_ happen after pulling the trigger.

Thanks to @victornpb for this script. I hope the above helps someone.

**Comment by ryz** [comment]
Over the last week I've deleted about ~200k messages on a server in various subchannels. 

I've used default values (100ms searchDelay, 1000ms deleteDelay) and pretty much just let the script run from morning until evening every day for 8-12 hours. 

I have kept discord usage in that time to a minimum - I didn't join any servers, add friends or similar. I only read a few messages and wrote 1-2 lines on accident via the mobile client because I forgot the script was running. 

So far I haven't been banned.

Thanks again for the script!

**Comment by 0xpr03** [comment]
Also had no problem with 500,1500ms delay + randomized 0-200 for every request. No server join/friend add etc in that time. Takes a while but gets the job done to have more privacy regarding older messages.

**Comment by hb0nes** [comment]
It's pathetic that Discord doesn't allow you to just bulk delete your DMs with one API call and starts banning people for it when they do it one by one as a bypass.

Don't hold back on implementing such a function only to start whining when people find their own way. 

Disgusting.

**Comment by Toby222** [comment]
> It's pathetic that Discord doesn't allow you to just bulk delete your DMs with one API call and starts banning people for it when they do it one by one as a bypass.
> 
> Don't hold back on implementing such a function only to start whining when people find their own way.
> 
> Disgusting.

You can request deletion of your messages, just have to specify which ones.
I don't recall where exactly, but one of the developers said you could theoretically just send them a list of every single messageID for your account. (You might get questioned why though.)
The reason they don't allow you to just delete everything on-demand/via API is that their business lies in messaging, so they  want to keep message history.

**Comment by dt-flo** [comment]
> 
> 
> It's pathetic that Discord doesn't allow you to just bulk delete your DMs with one API call and starts banning people for it when they do it one by one as a bypass.
> 
> Don't hold back on implementing such a function only to start whining when people find their own way.
> 
> Disgusting.

Your messages won't even get deleted if you delete your entire account. Never I will ever messages someone over a service that doesn't offer end-to-end encryption.

**Comment by user234683** [comment]
> You can request deletion of your messages, just have to specify which ones.

@Toby222 I did exactly that, for specific messages, and they completely ignored the email, even after sending follow-ups. So it is just PR as far as I can see. More likely explanation is that they don't want to allow bulk delete because it will limit future data-mining potential (there's no way Nitro is profitable). Money speaks louder than words.

**Comment by 0xpr03** [comment]
Increase randomization time max and search time, otherwise you'll get API rate limited for 6 seconds now.
Not gonna post values, I suspect discord devs frequent this.

**Comment by AnonymousMurid** [comment]
> Increase randomization time max and search time, otherwise you'll get API rate limited for 6 seconds now.
> Not gonna post values, I suspect discord devs frequent this.

I'm new to this platform and have little to no programming experience. Could you explain how to increase randomization time? I'm concerned about my account being disabled, however I have 175,000 DM messages I wish to delete. I would appreciate your help a lot. I understand if you don't wish to speak openly about this because I share your suspicions too. 

Hope to hear from you soon.

**Comment by 0xpr03** [comment]
Exchange the wait function line with the following:
`const wait = async ms => new Promise(done => setTimeout(done, ms+getRandomIntInclusive(MIN_VALUE,MAX_VALUE)));` and add the right value, pretty easy to figure out.
Also add the random function to the bottom
```javascript
function getRandomIntInclusive(min, max) {
  min = Math.ceil(min);
  max = Math.floor(max);
  return Math.floor(Math.random() * (max - min + 1) + min); //The maximum is inclusive and the minimum is inclusive
}
```

**Comment by Eratas** [comment]
200 ms search delay and 2000 ms delete delay is the sweet spot for me, otherwise I get rate limited.

Edit: This can cause some bots to crash and raise complaints, I suggest to increase both delays.

**Comment by aaa234243** [comment]
> Exchange the wait function line with the following:
> `const wait = async ms => new Promise(done => setTimeout(done, ms+getRandomIntInclusive(MIN_VALUE,MAX_VALUE)));` and add the right value, pretty easy to figure out.
> Also add the random function to the bottom
> 
> ```js
> function getRandomIntInclusive(min, max) {
>   min = Math.ceil(min);
>   max = Math.floor(max);
>   return Math.floor(Math.random() * (max - min + 1) + min); //The maximum is inclusive and the minimum is inclusive
> }
> ```

Does this stop the recent issue of being rate limited quickly into deleting, using the standard 100ms 1000ms search & import delay? @0xpr03

**Comment by 0xpr03** [comment]
> Does this stop the recent issue of being rate limited quickly into deleting, using the standard 100ms 1000ms search & import delay?

Not really, as you'll have to provide enough min/max delay for that to stop showing but, generally @Eratas values could work, just increase the min if you're having problems.

**Comment by Eratas** [comment]
And also make sure that channels aren't linked to other servers through bots, and that they're logging your delete requests, otherwise this will cause massive outage and probably result into a ban or something worst.

**Reply by Linerly** [reply]
AFAIK most moderation bots (e.g., Auttaja, Dyno) don't log message deletions after a while (as long there are new messages)

**Reply by djfjfofoc3** [reply]
> AFAIK most moderation bots (e.g., Auttaja, Dyno) don't log message deletions after a while (as long there are new messages)

However individual users using plugins such as Lighty's MessageLoggerV2, will still log messages. Despite these being against ToS seperately to modified discord clients, lots of people still use them.

**Comment by emmggi** [comment]
Hello guys.

Any new discoveries or techniques? Anyone got banned recently using this?

**Reply by Hitomiwari** [reply]
I have been using it for a while now deleted around like 70k messages in around 4 days and nothing happened to the account, will update you if I ever got banned.
Just a note I used it a lot few years ago before now and still the account wasn't banned.

**Comment by TryAnixx** [comment]
Is there now any code snippet that i can insert on my script to randomize the delays? or is this updated in the script?

**Comment by Slxzer** [comment]
Thank you for this. I need to delete like 3000 (3k) messages, i want to know the best deletion delay because i dont want my personal daily use account to get banned for this. I have a problem from only 1 server.

**Comment by KisamMeow** [comment]
I recently deleted like 400 messages and now Discord spams me with human verification every time I try to add a new friend or even accept a friend request and respond back.

**Reply by Hitomiwari** [reply]
That is a pretty normal reaction from Discord's api towards an IP that keeps sending requests constantly for a while, and that effect would last from half an hour to an hour once you stop deleting messages.

Then it should go back to normal, unless you are using a public VPN then that will happen most of the time.

**Reply by KisamMeow** [reply]
Well I deleted 50 messages on the 29th and Discord is still verifying that I'm human any time I try to send my first message or friend request, it's making me nervous. Lol

**Comment by VNRARA** [comment]


**Reply by Toby222** [reply]
Unrelated. Cloudflare just had an outage

**Reply by Animan8000** [reply]
It's very normal that Cloudflare may be unavailable for a short moment, therefore not being able to delete a message in that moment.

**Comment by LumiWasTaken** [comment]
![image](https://github.com/victornpb/undiscord/assets/49376128/6efad767-0e9d-4fb2-91eb-afa589257737)
Yup Got Disabled.

**Reply by yousef157** [reply]
Gonna tell us how? How many messages deleted, the delay set, other factors that might matter...

**Reply by Hitomiwari** [reply]
You must have been reported by whoever you were deleting their dm or if it was a server with lots of users then someone must have reported you.

**Reply by LumiWasTaken** [reply]
@yousef157 @Hitomiwari 

I've deleted about 940 messages inside a private server (for me and my bot)
Not sure how long it took, it was running in the background while i was playing final fantasy.

**Reply by LumiWasTaken** [reply]
i only used the browser extension, i didnt even use any better discord features

**Reply by Hitomiwari** [reply]
@LumiWasTaken it's weird I used undiscord on a lot of dms total deleting like 200-300k messages overall and the account is fine.

**Comment by ghost** [comment]
I’ve seen lots of people commenting on how if you get your account suspended due to deleting messages with your e-mail still attached to it it would stop you from creating new accounts. 
But what about payment methods/data? Will that be flagged in any way that if i pay nitro for another account and get mine suspended both will be flagged or suspended?
What about IP banning too? Any confirmations on if it happens?

**Reply by ghost** [reply]
also i have like 400k+ messages in a single server with an estimate of 400+ hours... how likely is that i delete all that at all?

**Reply by LumiWasTaken** [reply]
Might be the wrong discussion for that.

**Reply by Hitomiwari** [reply]
You can try, but no guarantee that you won't get banned.

**Reply by iArxic** [reply]
400+ hours is too huge, considering people get disabled for small deletes. if you do it please update us if you get disabled or not. 🙏

**Comment by VegaStarlake** [comment]
I'm unsure how this would be handled, so excuse me if I'm ignorant, but could Discord be looking for things like mouse movements similar to the captcha algorithms? My perspective may be based on old, macro-style scripts rather than what ever is going on here and I'm not well-versed enough to fully understand this repository. Does adjusting the order of the deleted messages and or also deleting reactions help reduce likelihood of detection? So far, I've had no issues with using this repository. I used it in a server and in dms, but it's been years and I am hesitant to use it again. Has anyone tried a basic screen-position mouse macro, and does anyone have any results from using that? I suspect that a randomized time/position (both screen and order) mouse macro could better avoid detection, but this is much less convenient of course and my logic could be completely off-base. Any thoughts?

**Reply by Hitomiwari** [reply]
I have been using it on actual dms regularly without mouse movement macros and whatsoever and it's all about not getting reported for it I think, I haven't been disabled once before for using it.

That been said I am not saying that you are safe to use it, whenever I use I am putting the possibility of losing my account, if you aren't ready to lose your account do not use it.

**Comment by regtable** [comment]
from what i see redact deletes messages at a rate of 25 messages a minute. this is according to discord tos they say. maybe staying on or under that limit is enough to save your acc

**Comment by phixxy** [comment]
I am deleting 400k+ messages. I will come back once done and post if I get banned or not. 2600ms delete delay. I suspect it will take me a few months to finish because I am not going to be doing it all at once.

**Reply by Sevenare1st** [reply]
How is it so far?

**Reply by phixxy** [reply]
I deleted 2000 messages with undiscord on the first day. Then I realized ~200k messages were in servers I have admin rights in. I made a simple discord cog/bot to delete all of those messages since that isn't against discord TOS. Once that is complete I will continue deleting dms with undiscord.

**Reply by Don-Cornelius-B** [reply]
What about DMs? and How is it till now?

**Comment by wavedevgit** [comment]
this is possibly because the script doesn't add `x-super-properties` header

**Comment by wavedevgit** [comment]
Made a discussion about the xsp header at #791
